### PR TITLE
Graph theory - minor improvements of Huneke's proof (8)

### PR DESCRIPTION
**Pull request by avekens**
General:
* Convention about "Organizing proofs" (conventions for lemmas) enhanced

Graph theory:

Neighbors/complete graphs:
* definitions of neighbors, universal vertices and complete graphs moved to the corresponding subsections
* theorems about neighbors revised: ~nbgrcl, ~nbgrnvtx0, ~nbgrel, ~nbgrisvtx, ~nbgrssvtx, ~nbgrnself2, ~nbgrssovtx, ~nbgrssvwo2, ~nbgrsym
* ~usgrnbnself2 marked as obsolete
* proofs of theorems about neighbors shortened: ~nbgr2vtx1edg, ~nbuhgr2vtx1edgblem, ~nbuhgr2vtx1edgb, ~nbgr0vtxlem, ~nbusgredgeu0, ~nbfusgrlevtxm1, ~nbfusgrlevtxm2, ~nb3grprlem2, nb3grprlem2, ~nbupgruvtxres
* proofs of theorems about complete graphs shortened: ~cplgr3v, ~cusgrexi

Closed walks as words:
* header comment of section "Closed walks as words" improved/enhanced
* theorems about closed walks as words revised: ~clwwlksn1
* new theorems about closed walks as words : ~loopclwwlksn1b, ~clwwlksn1loopb, ~0clwlkv

Friendship graphs:
* proofs of theorems about friendship graphs shortened: ~frgrnbnb, ~frgrncvvdeqlem1, ~frgrncvvdeqlem2, ~frgrncvvdeqlem3, ~frgrncvvdeqlem9, ~fusgreghash2wspv, ~numclwwlkovf2num, ~numclwlk1lem2foa, ~numclwlk1lem2fo
* theorems about friendship graphs revised: ~extwwlkfablem2, ~numclwwlkovf2exlem2, ~numclwwlkovf2
* comment for ~numclwwlkovf enhanced
* new theorems about friendship graphs: ~numclwwlkovf1, ~clwwlk1loop, ~clwwlk1nloop, ~clwwlk1, ~numclwwlkle1

### CODE REVIEW

File: set.mm [hunk 1]
```diff
@@ -408676,37 +408641,96 @@ function instead of the edges themselves (see also ~ nbgrval ).
...
   ${
-    $d e g n v $.  $d G g $.  $d N g $.
+    $d e g n v $.  $d G g $.  $d X g $.
     nbgrel.v $e |- V = ( Vtx ` G ) $.
-    $( There are no neighbors of a class which is not a vertex.  (Contributed
-       by Alexander van der Vekens, 12-Oct-2017.)  (Revised by AV,
-       26-Oct-2020.) $)
-    nbgrnvtx0 $p |- ( N e/ V -> ( G NeighbVtx N ) = (/) ) $=
+    $( There are no neighbors of a class ` X ` which is not a vertex of a graph
+       ` G ` .  (Contributed by Alexander van der Vekens, 12-Oct-2017.)
```
Comment (benjub) [inline set.mm:408647]: If a class ` X ` is not a vertex of a graph ` G ` , then it has no neighbors in ` G ` .

File: set.mm [hunk 2]
```diff
@@ -408676,37 +408641,96 @@ function instead of the edges themselves (see also ~ nbgrval ).
...
     nbgrel.e $e |- E = ( Edg ` G ) $.
-    $( Characterization of a neighbor of a vertex ` V ` in a graph ` G ` .
-       (Contributed by Alexander van der Vekens and Mario Carneiro,
-       9-Oct-2017.)  (Revised by AV, 26-Oct-2020.)  (Proof shortened by AV,
-       6-Jun-2021.) $)
-    nbgrel $p |- ( G e. W -> ( K e. ( G NeighbVtx N )
+    $( Characterization of a neighbor ` N ` of a vertex ` X ` in a graph
+       ` G ` .  (Contributed by Alexander van der Vekens and Mario Carneiro,
+       9-Oct-2017.)  (Revised by AV, 26-Oct-2020.)  (Revised by AV,
+       12-Feb-2022.) $)
```
Comment (benjub) [inline set.mm:408660]: the last two revisions are by the same person and are not explained, so maybe keeping one is enough ?

File: set.mm [hunk 3]
```diff
@@ -408676,37 +408641,96 @@ function instead of the edges themselves (see also ~ nbgrval ).
...
+      MUDVJVLECUEUFFCDEUGHUHUIUJVJVNVMVOVQWAMZMZMZWBVJVMVLWDVMVLWDUKVJVMVLDEJOZ
+      NZVSPZABQZJFEULUMZUNZKZWDVMVKWKDAJBCEFHIUOUPWLDWJKZWAMVOVQMZWAMWDWIWAJDWJ
+      WFDUQZWHVTABWOWGVRVSWFDEURUSUTVAWMWNWADFEVBRVOVQWASVCTVDVEWBVPWCMVMVOMZWC
+      MWEVPVQWAVFVPWPWCVOVMVHRVMVOWCSVGTVI $.
+  $}
+
+  ${
+    $d e g n v $.  $d G e g $.  $d K e g $.  $d N e $.  $d V e $.
+    nbgrisvtx.v $e |- V = ( Vtx ` G ) $.
+    $( Every neighbor of a class/vertex ` K ` is a vertex.  (Contributed by
```
Comment (benjub) [inline set.mm:408698]: Maybe:
class/vertex --> class (hence vertex)

File: set.mm [hunk 4]
```diff
@@ -408676,37 +408641,96 @@ function instead of the edges themselves (see also ~ nbgrval ).
...
+       Alexander van der Vekens, 12-Oct-2017.)  (Revised by AV, 26-Oct-2020.)
+       (Revised by AV, 12-Feb-2022.) $)
+    nbgrisvtx $p |- ( N e. ( G NeighbVtx K ) -> N e. V ) $=
+      ( ve cnbgr co wcel wa wne cpr cv wss cedg cfv wrex w3a eqid nbgrel simp1l
+      sylbi ) CABGHICDIZBDIZJCBKZBCLFMNFAOPZQZRUCFUFACDBEUFSTUCUDUEUGUAUB $.
+
+    $d G n $.  $d K n $.  $d V n $.
+    $( The neighbors of a vertex ` K ` in a graph are a subset of all vertices
```
Comment (benjub) [inline set.mm:408706]: are --> form

File: set.mm [hunk 5]
```diff
@@ -408806,73 +408831,72 @@ function instead of the edges themselves (see also ~ nbgrval ).
...
+      XDXSYHYDYIXDXSXPBYBPYHXDXPBXRYBXBXCVHVJXPYHBXCIVKZWNXCXOVLVMVNXDYDYABXQPY
+      IXDYABYCXQXBXCVOVJYAYIBXBHVKZWNXBXTVLVMVNVTVPUQVQVRXGXIYFSZXLXFUUFXDXFWMX
+      NXAYEEXECVLXFXAWQBXEWROZPZAXEPYEXFWTUUHAEXEXFUTXFWQBWSUUGEXEWRWAVJVSUUHXS
+      YDAXBXCUUEUUDWOXBKZWQXPBUUGXRUUIWRXQXEWOXBWBWCUUIWPXOWNWOXBDNWDWEVSWOXCKZ
+      WQYABUUGYCUUJWRYBXEWOXCWBWCUUJWPXTWNWOXCDNWDWEVSWFVNWKTTWGVRWHWIWJ $.
 
     $( Lemma for ~ nbuhgr2vtx1edgb .  This reverse direction of ~ nbgr2vtx1edg
        only holds for classes whose edges are subsets of the set of vertices
-       (hypergraphs!)  (Contributed by AV, 2-Nov-2020.) $)
+       (hypergraphs!)  (Contributed by AV, 2-Nov-2020.)  (Proof shortened by
```
Comment (benjub) [inline set.mm:408858]: final period. Or exclamation point after parenthesis ? Or no exclamation point ?

File: set.mm [hunk 6]
```diff
@@ -408952,116 +408976,126 @@ function instead of the edges themselves (see also ~ nbgrval ).
...
+       by AV, 3-Nov-2020.)  (Revised by AV, 12-Feb-2022.) $)
+    nbgrnself2 $p |- X e/ ( G NeighbVtx X ) $=
+      ( vv cvtx cfv wcel cnbgr co wnel cv wceq id oveq2 neleq12d eqid nbgrnself
+      vtoclri wn nbgrisvtx con3i df-nel sylibr pm2.61i ) BADEZFZBABGHZIZCJZAUHG
+      HZIUGCBUDUHBKZUHBUIUFUJLUHBAGMNCAUDUDOZPQUERBUFFZRUGULUEABBUDUKSTBUFUAUBU
+      C $.
+
+    $d V v $.
+    nbgrssovtx.v $e |- V = ( Vtx ` G ) $.
+    $( The neighbors of a vertex ` X ` are a subset of all vertices except the
```
Comment (benjub) [inline set.mm:409009]: are --> form

File: set.mm [hunk 7]
```diff
@@ -408952,116 +408976,126 @@ function instead of the edges themselves (see also ~ nbgrval ).
...
+       vertex itself.  Stronger version of ~ nbgrssvtx .  (Contributed by
+       Alexander van der Vekens, 13-Jul-2018.)  (Revised by AV, 3-Nov-2020.)
+       (Revised by AV, 12-Feb-2022.) $)
+    nbgrssovtx $p |- ( G NeighbVtx X ) C_ ( V \ { X } ) $=
+      ( vv cnbgr co csn cdif cv wcel nbgrisvtx wceq wn nbgrnself2 df-nel neleq1
+      wne wnel syl5bbr mpbiri necon2ai eldifsn sylanbrc ssriv ) EACFGZBCHIZEJZU
+      FKZUHBKUHCRUHUGKACUHBDLUIUHCUHCMZUINZCUFSZACOUKUHUFSUJULUHUFPUHCUFQTUAUBU
+      HBCUCUDUE $.
+
+    $( The neighbors of a vertex ` X ` are a subset of all vertices except the
```
Comment (benjub) [inline set.mm:409019]: are --> form

File: set.mm [hunk 8]
```diff
@@ -408952,116 +408976,126 @@ function instead of the edges themselves (see also ~ nbgrval ).
...
+       vertex itself and a vertex which is not a neighbor.  (Contributed by
```
Comment (benjub) [inline set.mm:409020]: vertex which is not --> class which is not

File: set.mm [hunk 9]
```diff
@@ -408952,116 +408976,126 @@ function instead of the edges themselves (see also ~ nbgrval ).
...
 
   ${
-    $d G e $.  $d K e $.  $d N e $.  $d W e $.
-    $( A vertex in a graph is a neighbor of a second vertex iff the second
-       vertex is a neighbor of the first vertex.  (Contributed by Alexander van
-       der Vekens, 12-Oct-2017.)  (Revised by AV, 27-Oct-2020.) $)
-    nbgrsym $p |- ( G e. W -> ( N e. ( G NeighbVtx K )
+    $d G e $.  $d K e $.  $d N e $.
+    $( A vertex ` N ` in a graph is a neighbor of a second vertex ` K ` iff the
+       second vertex ` K ` is a neighbor of the first vertex ` N ` .
```
Comment (benjub) [inline set.mm:409079]: Maybe add a second sentence:  In a graph, the neighborhood relation is symmetric.

File: set.mm [hunk 10]
```diff
@@ -409463,6 +409497,22 @@ is a bijection ( 1-1 onto function) in a simple graph.  (Contributed
...
+  $c UnivVtx $.
+
+  $( Extend class notation with the universal vertices (in a graph). $)
+  cuvtxa $a class UnivVtx $.
+
+  ${
+    $d g v n $.
+    $( Define the class of all universal vertices (in graphs).  A vertex is
+       called _universal_ if it is adjacent, i.e. connected by an edge, to all
+       other vertices (of the graph) resp. all other vertices are its
```
Comment (benjub) [inline set.mm:409509]: resp. --> , or equivalently, if all other....

File: set.mm [hunk 11]
```diff
@@ -409565,8 +409616,9 @@ is a bijection ( 1-1 onto function) in a simple graph.  (Contributed
       KULUM $.
 
     $d E v $.
-    $( If a graph/class has no edges, it has universal vertices if and only if
-       it has exactly one vertex.  (Contributed by AV, 30-Oct-2020.) $)
+    $( TODO-AV: antecedent ` G e. W ` can be removed!  If a graph/class has no
```
Comment (benjub) [inline set.mm:409619]: Ok, but try not to add too many TODOs and exclamation points in Main, or not for too long (the TODOs are ok in mathboxes).

File: set.mm [hunk 12]
```diff
@@ -409545,17 +409595,18 @@ is a bijection ( 1-1 onto function) in a simple graph.  (Contributed
...
+      MZUAUDJZCFVFUEZUBZNZAFOVEVFPZBMZQZBDRZCVINZAFOACEFGHUCVDVJVOAFVDVFFJZSZVG
+      VNCVIVGVEFJZVPSZVEVFUFZVFVEPZVLQZBDRZUGZVQVEVIJZSZVNBDEVEFVFHIUHWDVSVTSZW
+      CSZWFVNVSVTWCUIVNWCWFWHVMWBBDVKWAVLVEVFUJUKULWFWGWCWFVSVTVQVPWEVRVDVPUPVE
+      FVHUMUNWEVTVQVEFVFUOUQURUSUTTTVAVBVC $.
 
     $d E n $.  $d N e k n $.  $d W n $.
-    $( A universal vertex, i.e. an element of the set of all universal
-       vertices.  (Contributed by Alexander van der Vekens, 12-Oct-2017.) $)
+    $( TODO-AV: antecedent ` G e. W ` can be removed!  A universal vertex, i.e.
+       an element of the set of all universal vertices.  (Contributed by
```
Comment (benjub) [inline set.mm:409608]: Missing a verb ?

Comment (benjub) [reply set.mm:409608]: Or is it: "Characterization of a universal vertex" ?

File: set.mm [hunk 13]
```diff
@@ -409744,6 +409795,33 @@ is a bijection ( 1-1 onto function) in a simple graph.  (Contributed
...
+                                       v e. ( UnivVtx ` g ) } $.
+  $}
+
+  $( Define the class of all complete simple graphs.  A simple graph is called
+     _complete_ if every pair of distinct vertices is connected by a (unique)
+     edge, see definition in section 1.1 of [Diestel] p. 3.  In contrast, the
+     definition in section I.1 of [Bollobas] p. 3 is based on the size of
+     (finite) complete graphs, see ~ cusgrsize .  (Contributed by Alexander van
+     der Vekens, 12-Oct-2017.)  (Revised by AV, 24-Oct-2020.) $)
+  df-cusgr $a |- ComplUSGraph = { g e. USGraph | g e. ComplGraph } $.
```
Comment (benjub) [inline set.mm:409823]: possible to define it as an intersection ? ( USGraph i^i ComplGraph )  This saves a dummy variable

File: set.mm [hunk 14]
```diff
@@ -415270,9 +415349,9 @@ the walk is closed if v(n) = v0".
        represented by the following sequence: p(0) e(f(0)) p(1) e(f(1)) ...
        p(n-1) e(f(n-1)) p(n)=p(0).
 
-       Notice that by this definition, a single vertex is a closed walk of
-       length 0, see also ~ 0clwlk !  (Contributed by Alexander van der Vekens,
-       12-Mar-2018.)  (Revised by AV, 16-Feb-2021.) $)
+       Notice that by this definition, a single vertex can be considered as a
+       closed walk of length 0, see also ~ 0clwlk !  (Contributed by Alexander
```
Comment (benjub) [inline set.mm:415353]: exclamation point unnecessary

File: set.mm [hunk 15]
```diff
@@ -418964,17 +419043,22 @@ vertex v(0), so the number of walks of length n=0 starting with v=v(0)
...
-  "` v ` ` v `", which is a (closed) walk of length 1 (if there is an edge/loop
-  from ` v ` to ` v `).  Therefore, a closed walk corresponds to a closed walk
-  as word only for walks of length at least 1, see ~ clwlkclwwlk2 .  This is
-  also the reason for defining the set ` ClWWalksN ` of all closed walks of a
-  fixed length as words over the set of vertices as function over ` NN ` and
-  not ` NN0 `, because ` ( 0 ClWWalksN G ) = (/) ` (see ~ clwwlksn0 ) would
-  hold anyway.  In all other cases, however, a closed walk (of length at least
-  1) corresponds to a closed walk as word, at least in a simple pseudograph,
-  see ~ clwlkclwwlk2 .
+  To avoid to handle the degenerated case of representing a (closed) walk of
```
Comment (benjub) [inline set.mm:419046]: degenerate

File: set.mm [hunk 16]
```diff
@@ -419688,16 +419772,58 @@ the sequence p(0) p(1) ... p(n-1) of the vertices in a closed walk p(0)
...
-    clwwlksn1loop $p |- ( W e. ( 1 ClWWalksN G ) -> ( ( # ` W ) = 1
-                             /\ { ( W ` 0 ) , ( W ` 0 ) } e. ( Edg ` G ) ) ) $=
-      ( vi c1 cclwwlksn co wcel cvtx cfv cword chash wceq wa caddc cpr cedg cc0
-      cv cmin eqid cfzo wral clsw w3a clwwlknp adantr lsw1 preq1d eleq1d biimpa
-      simpr jca 3adant2 syl ) BDAEFGBAHIZJGZBKIDLZMZCRZBIUSDNFBIOAPIZGCQDDSFUAF
-      UBZBUCIZQBIZOZUTGZUDUQVCVCOZUTGZMZCUTADUOBUOTUTTUEURVEVHVAURVEMUQVGURUQVE
-      UPUQUKUFURVEVGURVDVFUTURVBVCVCUOBUGUHUIUJULUMUN $.
+    $( A closed walk of length 1 represented as word is a word consisting of 1
+       symbol representing a vertex connected by (at least) one edge with
+       itsself, i.e., by a loop.  (Contributed by AV, 24-Apr-2021.)  (Revised
```
Comment (benjub) [inline set.mm:419777]: itself
(and maybe: "connected to itself a (at least) one edge, that is, a loop")

**Review by benjub — COMMENTED** [review]
Review up to "clwwlksn1" included. A few comments so far.

File: set.mm (hunk 2 above)
Comment (avekens) [reply set.mm:408660]: I kept the first revision intentionally because this was the general revision of graph theory.

File: set.mm (hunk 11 above)
Comment (avekens) [reply set.mm:409619]: This theorem was already in main (and also used in proofs of theorems in main), therefore I had to tag it here. There are other TODOs in main, some still from Norm. Of course, such TODOs should be dealt with soon (as I will do with this one).

File: set.mm (hunk 13 above)
Comment (avekens) [reply set.mm:409823]: Good idea, I will earmark it for later PR (add TODO-AV ;-) ).

File: set.mm (hunk 1 above)
Comment (avekens) [reply set.mm:408647]: Done

File: set.mm (hunk 3 above)
Comment (avekens) [reply set.mm:408698]: Done

File: set.mm (hunk 4 above)
Comment (avekens) [reply set.mm:408706]: Done

File: set.mm (hunk 5 above)
Comment (avekens) [reply set.mm:408858]: Done

File: set.mm (hunk 6 above)
Comment (avekens) [reply set.mm:409009]: Done

File: set.mm (hunk 7 above)
Comment (avekens) [reply set.mm:409019]: Done

File: set.mm (hunk 8 above)
Comment (avekens) [reply set.mm:409020]: Done

File: set.mm (hunk 9 above)
Comment (avekens) [reply set.mm:409079]: Done

File: set.mm (hunk 10 above)
Comment (avekens) [reply set.mm:409509]: Done

File: set.mm (hunk 12 above)
Comment (avekens) [reply set.mm:409608]: Done

File: set.mm (hunk 14 above)
Comment (avekens) [reply set.mm:415353]: Done

File: set.mm (hunk 15 above)
Comment (avekens) [reply set.mm:419046]: Done

File: set.mm (hunk 16 above)
Comment (avekens) [reply set.mm:419777]: Done

File: set.mm [hunk 17]
```diff
@@ -421310,6 +421439,17 @@ corresponds to the size of the set of closed walks of length n (in an
       isclwlk ancom bitri hash0 eqcomi fveq2i biantrur 3bitr4g ) BDFZGAHIJHZAHK
       ZIABLHMZNZUKGGOPCAQZNIABRHMZUNUIULUNUKADBCESTUOULUKNUMAIBUAULUKUBUCUKUNGU
       JAUJGUDUEUFUGUH $.
+
+    $( Any vertex (more precisely, a pair of an empty set (of edges) and a
+       singleton function to this vertex) determines a closed walk of length 0.
+       (Contributed by AV, 11-Feb-2022.) $)
+    0clwlkv $p |- ( ( X e. V /\ F = (/) /\ P : { 0 } --> { X } )
+                    -> F ( ClWalks ` G ) P ) $=
+      ( wcel c0 wceq cc0 csn wf w3a cclwlks cfv wbr cfz 3ad2ant1 wb cvv co fssd
```
Comment (benjub) [inline set.mm]: maybe `|- ( ( X e. V /\ P : { 0 } --> { X } ) -> (/)  ( ClWalks `` G ) P ) $= ` ?

File: set.mm [hunk 18]
```diff
@@ -425498,66 +425636,60 @@ that the number n of vertices in G is exactly k(k-1)+1.".  Variant of
...
        22-Sep-2018.)  (Revised by AV, 27-Jan-2022.) $)
-    numclwwlkovf2exlem2 $p |- ( ( ( ( G e. USGraph /\ X e. V
-                                      /\ N e. ( ZZ>= ` 3 ) )
+    numclwwlkovf2exlem2 $p |- ( ( ( ( X e. V /\ Y e. V /\ N e. ( ZZ>= ` 3 ) )
                    /\ ( ( W e. Word V /\ A. i e. ( 0 ..^ ( ( # ` W ) - 1 ) )
                                          { ( W ` i ) , ( W ` ( i + 1 ) ) } e. E
                    /\ { ( lastS ` W ) , ( W ` 0 ) } e. E )
                    /\ ( # ` W ) = ( N - 2 ) /\ ( W ` 0 ) = X ) )
-                   /\ Y e. ( G NeighbVtx X ) )
+                   /\ { X , Y } e. E )
```
Comment (benjub) [inline set.mm]: By the way: I presume the equivalence of these two lines (i.e., under the required graphness and sethood antecedents,  Y e. ( G NeighbVtx X ) <-> { X , Y } e. E ) is already in Main ? As well as  X e. ( G NeighbVtx Y ) <-> Y e. ( G NeighbVtx X ) ?

File: set.mm [hunk 19]
```diff
@@ -425567,8 +425699,16 @@ that the number n of vertices in G is exactly k(k-1)+1.".  Variant of
...
        from v to v of length n" according to definition 5 in [Huneke] p. 2.
-       (Contributed by Alexander van der Vekens, 14-Sep-2018.)  (Revised by AV,
-       28-May-2021.) $)
+
+       ` ( X F N ) ` is not defined for ` N = 0 ` , since closed walks of
+       length 0 cannot be represented by a sequence (word) of vertices, see
+       ~ clwwlksn0 . ` ( X F N ) = (/) ` could have been explicitly defined,
+       but the operation value for operators for which the operation is not
+       defined is already ` (/) ` by definition (see ~ ndmovg ), therefore such
+       an extension of the definition of ` F ` is not necessary.  Nevertheless,
```
Comment (benjub) [inline set.mm]: Nevertheless --> In any case
?

**Review by benjub — APPROVED** [review]
Finished my review, and approved, with a few suggestions that can be dealt with now or later. In other words: ready to merge.

More generally: I find the whole exposition (not just this PR) a bit hard to follow, and I'm not sure if this is me not putting enough time to enter into this entangled web of slightly differing definitions, if this is the field of graph theory which makes it this way because (and it is true) there actually are many useful definitions which differ slightly (e.g., paths (open, semi-open, etc.), arcs, walks, closed walks, loops) and there is no one universally more natural than the others: the diversity of applications of graph theory makes one notion sometimes more natural than the others, but it varies... Or maybe the set.mm treatment can be streamlined a bit. Maybe @avekens, @tirix and @digama0 have some thoughts to share about this.

File: set.mm [hunk 20]
```diff
@@ -409744,6 +409796,35 @@ is a bijection ( 1-1 onto function) in a simple graph.  (Contributed
...
+  ccusgr $a class ComplUSGraph $.
+
+  ${
+    $d g v $.
+    $( Define the class of all complete "graphs".  A class/graph is called
+       _complete_ if every pair of distinct vertices is connected by an edge,
+       i.e. each vertex has all other vertices as neighbors.  (Contributed by
+       AV, 24-Oct-2020.) $)
+    df-cplgr $a |- ComplGraph = { g | A. v e. ( Vtx ` g )
+                                       v e. ( UnivVtx ` g ) } $.
```
Comment (tirix) [inline set.mm]: I suppose you could also save a bound variable here by writing
```metamath
ComplGraph = { g | ( Vtx ` g ) C_ ( UnivVtx ` g ) }
```
This is slightly cryptic, but in general I think we favor short definition over expressive definitions, and then express the desired property in a more human readable way in a subsequent theorem.

File: set.mm [hunk 21]
```diff
@@ -415270,9 +415352,9 @@ the walk is closed if v(n) = v0".
...
        p(n-1) e(f(n-1)) p(n)=p(0).
 
-       Notice that by this definition, a single vertex is a closed walk of
-       length 0, see also ~ 0clwlk !  (Contributed by Alexander van der Vekens,
-       12-Mar-2018.)  (Revised by AV, 16-Feb-2021.) $)
+       Notice that by this definition, a single vertex can be considered as a
+       closed walk of length 0, see also ~ 0clwlk .  (Contributed by Alexander
+       van der Vekens, 12-Mar-2018.)  (Revised by AV, 16-Feb-2021.) $)
     df-clwlks $a |- ClWalks = ( g e. _V |-> { <. f , p >. |
                  ( f ( Walks ` g ) p /\ ( p ` 0 ) = ( p ` ( # ` f ) ) ) } ) $.
```
Comment (tirix) [inline set.mm]: Maybe some confusion come from the different definitions in the literature.
In the comment, you write that the definition of Huneke for a closed walk is a sequence of vertices, however the definition we use for walks is the one from Bollobas, which is an alternating sequence of vertices and edges... therefore the different properties, I assume?

File: set.mm [hunk 22]
```diff
@@ -425567,8 +425699,16 @@ that the number n of vertices in G is exactly k(k-1)+1.".  Variant of
     $( Value of operation ` F ` , mapping a vertex ` v ` and a positive integer
        ` n ` to the "(For a fixed vertex v, let f(n) be the number of) walks
        from v to v of length n" according to definition 5 in [Huneke] p. 2.
-       (Contributed by Alexander van der Vekens, 14-Sep-2018.)  (Revised by AV,
-       28-May-2021.) $)
+
```
Comment (tirix) [inline set.mm]: Why not simply write:
> Value of operation ` F ` , mapping a vertex ` v ` and a positive integer ` n ` to the number of walks from `v` to `v` of length `n`" according to definition 5 in [Huneke] p. 2.

(sorry this is not even in your change set!)

Comment (tirix) [inline set.mm]: Why not simply write:
> Value of operation ` F ` , mapping a vertex ` v ` and a positive integer ` n ` to the number of walks from `v` to `v` of length `n`" according to definition 5 in [Huneke] p. 2.

(sorry this is not even in your change set!)

File: set.mm [hunk 23]
```diff
@@ -425601,82 +425741,145 @@ that the number n of vertices in G is exactly k(k-1)+1.".  Variant of
...
-      chash c1 cword cn simpr numclwwlkovf sylancl cvtx cedg wb clwwlksn2 anass
-      2nn a1i df-3an ancom eqcomi wrdeqi eleq2i anbi12i anbi2i syl6bb rabbidva2
-      3bitr4i eqtrd ) FUALZHGLZMZHNEOZUBAUCZPZHQZANFUDOZRZVPUGPNQZVQUHVPPUEZDLZ
-      VRUFZAGUIZRVNVMNUJLVOVTQVLVMUKUSABCEFNGHIULUMVNVRWDAVSWEVNVPVSLZVRMZWAVPF
-      UNPZUIZLZWBFUOPZLZUFZVRMZVPWELZWDMZWGWNUPVNWFWMVRFVPUQSUTWOWAWCMZMZVRMWOW
-      QVRMZMWNWPWOWQVRURWMWRVRWMWOWAMZWCMZWRWMWAWJMZWLMXAWAWJWLVAXBWTWLWCXBWJWA
-      MWTWAWJVBWJWOWAWIWEVPWHGGWHJVCVDVESTWKDWBDWKKVCVEVFTWOWAWCURTSWDWSWOWAWCV
-      RVAVGVJVHVIVK $.
+    $( Value of operation ` F ` for the second operand being 1.  (Contributed
+       by AV, 11-Feb-2021.) $)
```
Comment (tirix) [inline set.mm]: Maybe "set of walks of length one starting from vertex `v`" would be clearer?

Comment (tirix) [inline set.mm]: Maybe "set of closed walks of length one starting from vertex `v`" would be clearer?

File: set.mm (hunk 22 above)
Comment (tirix) [inline set.mm]: It seems this defines the _set_ of walks, not the number of walks.
Also, it seems this defines the set of _closed_ walks, not just walks.

**Review by tirix — APPROVED** [review]

**Comment by tirix** [comment]
> I find the whole exposition (not just this PR) a bit hard to follow, and I'm not sure if this is me not putting enough time to enter into this entangled web of slightly differing definitions, if this is the field of graph theory which makes it this way because (and it is true) there actually are many useful definitions which differ slightly (e.g., paths (open, semi-open, etc.), arcs, walks, closed walks, loops) and there is no one universally more natural than the others: the diversity of applications of graph theory makes one notion sometimes more natural than the others, but it varies... Or maybe the set.mm treatment can be streamlined a bit. Maybe @avekens, @tirix and @digama0 have some thoughts to share about this.

I think you're very right. Depending on the context, it might be better to define a walk as a sequence of vertices, including or not the last vertex, a sequence of edges, or even an alternating sequence of vertices and edges.

And then there are distinctions between walks, trails, paths, cycles, circuits...

For formalization, it would probably be easier to follow if it was possible to stick to only one (the most generic) definition, and if all those concepts were specific cases, so that we would always deal with the same kind of objects (like what we have for structures). However then we lose the ability to use "tricks" like cyclic shifts, and we probably would not stick to the litterature anymore.

Maybe something which would also help is a visualisation of some of the most complex cases (it would also help a lot in elementary geometry!). If I find the time, I'd like to draw the hierarchy of graph families like I did for the algebraic structures: this would help a little bit.

File: set.mm (hunk 18 above)
Comment (avekens) [reply set.mm]: Y e. ( G NeighbVtx X ) <-> { X , Y } e. E ) : ~nbupgrel
X e. ( G NeighbVtx Y ) <-> Y e. ( G NeighbVtx X ) : ~nbgrsym

File: set.mm (hunk 17 above)
Comment (avekens) [reply set.mm]: Since this theorem is for demonstration purposes only (not (yet) used in a proof), I want to keep this versio, because a walk, trail, path, etc. is represented by a pair <. F , P >. of (indexed) edges F and (indexed) vertices P.

File: set.mm (hunk 19 above)
Comment (avekens) [reply set.mm]: "In any case, the case ..." sounds a little bit clumsy.

File: set.mm (hunk 20 above)
Comment (avekens) [reply set.mm]: Good idea! I will change it in my next PR accordingly.
Since ( UnivVtx ` g ) C_ ( Vtx ` g ) holds anyway, we also can define ComplGraph = { g | ( UnivVtx ` g ) =  ( Vtx ` g ) }.
Which variant would you prefer?

**Comment by avekens** [comment]
@benjub and @tirix : thank you very much for your thorough review and your valuable remarks. I think for a general discussion, this PR is not the right place. Maybe you can open an issue on Github or a thread in the Google groop for this. Therefore, I give only some short answers/hints in the following.
 
> > I find the whole exposition (not just this PR) a bit hard to follow, and I'm not sure if this is me not putting enough time to enter into this entangled web of slightly differing definitions, if this is the field of graph theory which makes it this way because (and it is true) there actually are many useful definitions which differ slightly (e.g., paths (open, semi-open, etc.), arcs, walks, closed walks, loops) and there is no one universally more natural than the others: the diversity of applications of graph theory makes one notion sometimes more natural than the others, but it varies... Or maybe the set.mm treatment can be streamlined a bit. Maybe @avekens, @tirix and @digama0 have some thoughts to share about this.

In general, the definitions are already streamlined in set.mm (my work during the last 5 years, based on the definitions and suggestions of @digama0. The (mainly) two variants of defining walks, path, etc. are based on two purposes: The more general variant using alternating sequences of edges and vertices can be used for (almost) all kinds of graphs (hypergraphs, multigraphs, etc.) and is required/approriate to proof the Koengisberg Bridge problem, whereas the other variant using sequences of vertices only are meaningful for simple (pseudo)graphs only and are required/approriate to proof the Friendship theorem.
> 
> I think you're very right. Depending on the context, it might be better to define a walk as a sequence of vertices, including or not the last vertex, a sequence of edges, or even an alternating sequence of vertices and edges.

Including or excuding the last vertex is a question about the representation of **closed** walks/paths/etc. only.
> 
> And then there are distinctions between walks, trails, paths, cycles, circuits...

There is a lot of material/information about the distinction  in the comments of the definitions, but also in the header comments of some subsections. Maybe I should give a general overview in the header comment of section "Walks, paths and cycles".
> 
> For formalization, it would probably be easier to follow if it was possible to stick to only one (the most generic) definition, and if all those concepts were specific cases, so that we would always deal with the same kind of objects (like what we have for structures). However then we lose the ability to use "tricks" like cyclic shifts, and we probably would not stick to the litterature anymore.

As explained above, to prove the friendship theorem with the general definitions for walks/paths/etc. would be very hard an much more difficult to understand.
> 
> Maybe something which would also help is a visualisation of some of the most complex cases (it would also help a lot in elementary geometry!). If I find the time, I'd like to draw the hierarchy of graph families like I did for the algebraic structures: this would help a little bit.

The graph hierarchy is textually availabe in the header comment of section "Undirected graphs". It would be great if you can visualize it.

File: set.mm (hunk 21 above)
Comment (avekens) [reply set.mm]: See my response to the general comment.

File: set.mm (hunk 22 above)
Comment (avekens) [reply set.mm]: Yes, to admix the citation into the explanation is really confusing. I will change this with my next PR.

**Comment by avekens** [comment]
@tirix Sorry, I am not able to comment your last remark (I will consider it in my next PR). 

@tirix @benjub Since it becomes more and more complicated to follow the remarks in this PR, I want to ask you to accept and merge it as soon as possible.

File: set.mm (hunk 20 above)
Comment (tirix) [reply set.mm]: I think I would personally prefer the second,
```
ComplGraph = { g | ( UnivVtx ` g ) = ( Vtx ` g ) }
```

File: set.mm (hunk 22 above)
Comment (tirix) [reply set.mm]: Oh! Actually I did not even think it could be a verbatim citation!
Even so, I still think it's less confusing to restate the sentence.

**Comment by tirix** [comment]
> @tirix Sorry, I am not able to comment your last remark (I will consider it in my next PR).

Strange, I think there was some kind of network error when I posted it. Anyway, no problem.

> @tirix @benjub Since it becomes more and more complicated to follow the remarks in this PR, I want to ask you to accept and merge it as soon as possible.

I'm fine with merging, and I think @benjub too.

File: set.mm (hunk 20 above)
Comment (benjub) [reply set.mm]: @tirix : that's not cryptic... *this* is cryptic:
```
ComplGraph = ( Graph i^i ( ( `' ( Vtx X2 UnivVtx ) ) " _I ) )
```
where X2 denotes the "range cartesian product". It sounds better than it reads: "the inverse image of the diagonal under the map (Vtx, UnivVtx)".  We see the three-level progression: two dummy variables, one dummy variable, no dummy variables: increasingly compact, terse and economical, at the price of being a bit more cryptic. One could say that the latter definition is "à la @sctfn" !  Maybe use it ? (which would bring the range cartesian product to Main...)
